File: clients/JavaScript/model/client-message.js
const DebugCommand = require('./debug-command');
const Action = require('./action');
class ClientMessage {
    static async readFrom(stream) {
        let tag = await stream.readInt();
        if (tag == DebugMessage.TAG) {
            return await DebugMessage.readFrom(stream);
        }
        if (tag == ActionMessage.TAG) {
            return await ActionMessage.readFrom(stream);
        }
        if (tag == DebugUpdateDone.TAG) {
            return await DebugUpdateDone.readFrom(stream);
        }
        if (tag == RequestDebugState.TAG) {
            return await RequestDebugState.readFrom(stream);
        }
        throw new Error("Unexpected tag value");
    }
}

class DebugMessage extends ClientMessage {
    constructor(command) {
        super();
        this.command = command;
    }
    static async readFrom(stream) {
        let command;
        command = await DebugCommand.readFrom(stream);
        return new DebugMessage(command);
    }
    async writeTo(stream) {
        await stream.writeInt(DebugMessage.TAG);
        let command = this.command;
        await command.writeTo(stream);
    }
}
ClientMessage.DebugMessage = DebugMessage;
DebugMessage.TAG = 0;
class ActionMessage extends ClientMessage {
    constructor(action) {
        super();
        this.action = action;
    }
    static async readFrom(stream) {
        let action;
        action = await Action.readFrom(stream);
        return new ActionMessage(action);
    }
    async writeTo(stream) {
        await stream.writeInt(ActionMessage.TAG);
        let action = this.action;
        await action.writeTo(stream);
    }
}
ClientMessage.ActionMessage = ActionMessage;
ActionMessage.TAG = 1;
class DebugUpdateDone extends ClientMessage {
    constructor() {
        super();
    }
    static async readFrom(stream) {
        return new DebugUpdateDone();
    }
    async writeTo(stream) {
        await stream.writeInt(DebugUpdateDone.TAG);
    }
}
ClientMessage.DebugUpdateDone = DebugUpdateDone;
DebugUpdateDone.TAG = 2;
class RequestDebugState extends ClientMessage {
    constructor() {
        super();
    }
    static async readFrom(stream) {
        return new RequestDebugState();
    }
    async writeTo(stream) {
        await stream.writeInt(RequestDebugState.TAG);
    }
}
ClientMessage.RequestDebugState = RequestDebugState;
RequestDebugState.TAG = 3;
module.exports = ClientMessage;

File: clients/Cpp/model/ClientMessage.cpp
#include "ClientMessage.hpp"
#include <stdexcept>

ClientMessage::DebugMessage::DebugMessage() { }
ClientMessage::DebugMessage::DebugMessage(std::shared_ptr<DebugCommand> command) : command(command) { }
ClientMessage::DebugMessage ClientMessage::DebugMessage::readFrom(InputStream& stream) {
    ClientMessage::DebugMessage result;
    result.command = DebugCommand::readFrom(stream);
    return result;
}
void ClientMessage::DebugMessage::writeTo(OutputStream& stream) const {
    stream.write(TAG);
    command->writeTo(stream);
}

ClientMessage::ActionMessage::ActionMessage() { }
ClientMessage::ActionMessage::ActionMessage(Action action) : action(action) { }
ClientMessage::ActionMessage ClientMessage::ActionMessage::readFrom(InputStream& stream) {
    ClientMessage::ActionMessage result;
    result.action = Action::readFrom(stream);
    return result;
}
void ClientMessage::ActionMessage::writeTo(OutputStream& stream) const {
    stream.write(TAG);
    action.writeTo(stream);
}

ClientMessage::DebugUpdateDone::DebugUpdateDone() { }
ClientMessage::DebugUpdateDone ClientMessage::DebugUpdateDone::readFrom(InputStream& stream) {
    ClientMessage::DebugUpdateDone result;
    return result;
}
void ClientMessage::DebugUpdateDone::writeTo(OutputStream& stream) const {
    stream.write(TAG);
}

ClientMessage::RequestDebugState::RequestDebugState() { }
ClientMessage::RequestDebugState ClientMessage::RequestDebugState::readFrom(InputStream& stream) {
    ClientMessage::RequestDebugState result;
    return result;
}
void ClientMessage::RequestDebugState::writeTo(OutputStream& stream) const {
    stream.write(TAG);
}
std::shared_ptr<ClientMessage> ClientMessage::readFrom(InputStream& stream) {
    switch (stream.readInt()) {
    case 0:
        return std::shared_ptr<ClientMessage::DebugMessage>(new ClientMessage::DebugMessage(ClientMessage::DebugMessage::readFrom(stream)));
    case 1:
        return std::shared_ptr<ClientMessage::ActionMessage>(new ClientMessage::ActionMessage(ClientMessage::ActionMessage::readFrom(stream)));
    case 2:
        return std::shared_ptr<ClientMessage::DebugUpdateDone>(new ClientMessage::DebugUpdateDone(ClientMessage::DebugUpdateDone::readFrom(stream)));
    case 3:
        return std::shared_ptr<ClientMessage::RequestDebugState>(new ClientMessage::RequestDebugState(ClientMessage::RequestDebugState::readFrom(stream)));
    default:
        throw std::runtime_error("Unexpected tag value");
    }
};
